cpp/src/arrow/c/abi.h

```diff
@@ -60,8 +63,21 @@ struct ArrowArray {
   void* private_data;
 };
 
+#ifdef __cplusplus
```

Comment:
I would put the `__cplusplus` guards outside of the canonical C data guards, so that it's a simple copy/paste job from the spec.